Fix integration tests

pablochacin opened the pull request:
## What does this PR change?
This PR updates the integration tests so they check for the actual copy of the certificates to the intended locations.  

It also introduces enhancements for testing against an existing K8S cluster like minikube or kind.

The tests doesn't work against the kubebuilder provided test environment, as they rely on access to nodes' metadata, which is not available in that environment. 

## Documentation

Added section in docs/devel.md about how to test against an existing cluster

## Test coverage

- Integration test were modify.  

- [ ] **DONE**

## Links

- [ ] **DONE**

MalloZup commented:
i think we need to `integrate` this in the `tests/e2e.sh` script runner

inercia commented on pkg/test/certificates/foo.crt:
```diff
@@ -0,0 +1,35 @@
+-----BEGIN CERTIFICATE-----
```
Theere is already a `pkg/test/assets/assets.go` file with a certificate inside. Couldn't we unify both things in only one certificate?

pablochacin replied:
The files are not longer used, as they were required from different tests and the path for accessing them was relative to the test's directory. Therefore the certs were embedded in a go source file in the pkg/test/asset package.

inercia commented on pkg/test/certificates/foo.key:
```diff
@@ -0,0 +1,52 @@
+-----BEGIN PRIVATE KEY-----
```
It seems that the certifcicate generation is an important part of these tests. I was wondering wether it wound't be better to add some functions for [generating certificates](https://golang.org/src/crypto/tls/generate_cert.go) dynamically (so we could skip all these fixtures in different formats)...

pablochacin replied:
That could be nice enhancement.  But I think it is not critical now.

pablochacin replied:
In a second thought, the idea of having the certificates as  fixtures is to ensure that the same certificate is retrieved each time. I'm thinking in the situation when we move some common setup for tests to a setup function that is reused in multiple tests.If in the test I need to check the same secret, how to ensure this if it is generated automatically?

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -59,29 +61,69 @@ func TestReconcile(t *testing.T) {
 	g.Expect(err).NotTo(gomega.HaveOccurred())
-	defer c.Delete(context.TODO(), instance)
+
+	defer func() {
```
this defer is to big. in a `defer` call, if we have error we cannot handle/catch them

pablochacin replied:
This defer cleans up the resources for this particular test. I'm using defer because there are many points where the test can exit. The only alternative I see is to refactor the whole test as follows:
`
func setupX(){
//setup test X
}

func cleanupX(){
//tear down test X
}

doX(){
 //actually execute test X
}

func testX(){
setupX()
doX()
cleanupX()
}
`

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -59,29 +61,69 @@ func TestReconcile(t *testing.T) {
+			//Ensure the finalizers are removed to prevent delete to be
+			//deadlocked by failed operator
+			created.ObjectMeta.Finalizers = []string{}
+			c.Update(context.TODO(), created)
```
we are not checking the error here.

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -59,29 +61,69 @@ func TestReconcile(t *testing.T) {
+		created := &kubicv1beta1.Registry{}
+
+		err0 := c.Get(context.TODO(), registryName, created)
+		if err0 == nil {
```
we should also check if the err != nil ( but in defer is not possible so we need to move out)

MalloZup replied:
also `err0` is not usual. we should call it `err`

pablochacin commented:
@MalloZup @inercia @ereslibre I have addressed the issues you pointed out in the review and on the chat. The PR is ready for another round of reviews. Thanks.

MalloZup commented on pkg/controller/registry/registry_controller_suite_test.go:
```diff
@@ -62,21 +65,38 @@ func TestMain(m *testing.M) {
+}
+
+func  SetupTestManager(t *testing.T) manager.Manager {
+
```
we can remove this `space`

MalloZup commented on pkg/controller/registry/registry_controller_suite_test.go:
```diff
@@ -62,21 +65,38 @@ func TestMain(m *testing.M) {
 // writes the request to requests after Reconcile is finished.
-func SetupTestReconcile(inner reconcile.Reconciler) (reconcile.Reconciler, chan reconcile.Request) {
+func SetupTestReconciler(inner reconcile.Reconciler) (reconcile.Reconciler, chan reconcile.Request) {
+
```
afaik we could remove empty space but just a nitpick

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -18,70 +18,298 @@
+		err  = c.Update(context.TODO(), registry)
+		if err != nil {
+		  t.Logf("Error updating registry %v", err)
+		} else {
```
afaik we can write like this
```golang
if err != nil {
   t.Logf("Error updating registry %v", err)
} 
err = c.Delete(context.TODO(), registry)
if err != nil {
  t.Logf("Error deleting registry %v", err)
}
```

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -18,70 +18,298 @@
 func TestReconcile(t *testing.T) {
 
-	test.SkipUnlessIntegrationTesting(t)
+	var err error
```
we can remove this and by line 276 we can use `err = addRegController(mgr,newRegistryReconcilier(mgr))`

pablochacin replied:
As the variable is used in many places, when adding new error check I found I have to change when it was first declared. I think that this makes clear it is a global variable for the whole function.

MalloZup replied:
In golang is a tacit consensus to reuse same var err everywhere without to declare it like that. It is a common idiom, checkout other projects etc

pablochacin commented on pkg/controller/registry/registry_controller_suite_test.go:
```diff
@@ -62,21 +65,38 @@ func TestMain(m *testing.M) {
-		g.Expect(mgr.Start(stop)).NotTo(gomega.HaveOccurred())
+		err := mgr.Start(stop)
+		if err != nil {
+			t.Errorf("Failed to start Manager: %v", err)
```
Only for reporting. I haven't found how to make the test fail in this situation which should be the best option, because without the manager, the test won't run.

pablochacin replied:
To clarify. The call to `mgr.Start(stop)` will return an error only if there's an error, otherwise, it will block until the `stop` channel is closed. Therefore, trying to return the error in a channel:
```
func StartTestManager(t *testing.T, mgr manager.Manager) (chan struct{}, chan error) {
        stop := make(chan struct{})                                                       
        errCh := make(chan error)                                                         
        go func() { 
                err := mgr.Start(stop)
                errCh <- err                                                              
        }()                                                                               
        return stop, errCh                                                                
}                                                                                         
```
Won't work because in  `TestReconcile` the following code blocks:
```
stop, errorCh := StartTestManager(t, mgr)
err = <- erroCh
if err != nil {
     t.Fatalf("Error starting manager %v", err)
}
defer close(stop)
```

MalloZup replied:
ok lets' resolve this ths @pablochacin

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -18,70 +18,290 @@
+	numNodes := int32(0)
+	var err error
+	wait.Poll(100*time.Millisecond, timeout, func() (bool, error) {
+		err = c.Get(context.TODO(), types.NamespacedName{Name: jobName, Namespace: metav1.NamespaceSystem}, job)
```
```golang
err = c.Get(context.TODO(), types.NamespacedName{Name: jobName, Namespace: metav1.NamespaceSystem}, job)
		if err != nil && !apierrors.IsNotFound(err) {
			return false, err
		}
	numNodes = job.Status.Succeeded
```

MalloZup replied:
i think we could simplify like this, removing if/else

pablochacin replied:
The code was originally like that, but if `apierrors.IsNotFound(err)` is true then accessing job.Status.Succeeded is not safe.

MalloZup replied:
ok we can refactor later then .

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -18,70 +18,290 @@
+		} else if !apierrors.IsNotFound(err) {
+			return false, err
+		}
+		err = nil
```
with the above logic we could remove this

MalloZup commented on pkg/controller/registry/registry_controller_test.go:
```diff
@@ -18,70 +18,290 @@
+	return err
+}
+
+func waitRegistryUpdates(c client.Client, registryName string, expected int32, timeout time.Duration) (int32, error) {
```
```golang
func waitRegistryUpdates(c client.Client, registryName string, expected int32, timeout time.Duration) (int32, error) {
	registry := &kubicv1beta1.Registry{}
	numNodes := int32(-1)
	wait.Poll(100*time.Millisecond, timeout, func() (bool, error) {
		err := c.Get(context.TODO(), types.NamespacedName{Name: registryName}, registry)
		if err != nil &&  !apierrors.IsNotFound(err) {
			return false, err
		}
                numNodes = int32(registry.Status.Certificate.NumNodes)
		return (numNodes == expected), nil
	})
	return numNodes, nil
}
```

MalloZup replied:
i think we could maybe simplify like this, i didn't tried

pablochacin replied:
The code you propose brakes the intended behavior. The intention is to return the `err` catched inside the function passed to `Poll`  from the `waitRegistryUpdates`. In the code you propose,  `err` is missing.
